content/post/2019-06-18.md

```diff
@@ -0,0 +1,256 @@
+
+As Gephi said:
+
+> GDF is built like a database table or a coma separated file (CSV). It supports attributes to both nodes and edges. A standard file is divided in two sections, one for nodes and one for edges. Each section has a header line, which basically is the column title. Each element (i.e. node or edge) is on a line and values are separated by coma. The GDF format is therefore very easy to read and can be easily converted from CSV.
+
+This is a basic example of a GDF file, in a first part it defines the nodes and in a second the connections:
+
+```
+nodedef>name VARCHAR,label VARCHAR
+s1,Site number 1
+s2,Site number 2
+s3,Site number 3
+edgedef>node1 VARCHAR,node2 VARCHAR
+s1,s2
+s2,s3
+s3,s2
+s3,s1
+```
+
+The steps to extract relationships from a CSV file and generate a file in GDF format are:
+
+1. Define the parameters for the extraction of related data
+2. Define the data structures for the transformation
+3. Import the data from a file in CSV format and store the data in the structures
+4. Sort data according to connections
+5. Prepare the data for the GDF format
+6. Generate the file in GDF format
+
+## Define the parameters for the extraction of related data
+
+A set of parameters are defined so this converter can be adapted to different cases. We must specify what information will be taken from the file in CSV format, where we will leave the result and the graph type.
+
+```r
+library("readr")
+library("hash")
+library("stringr")
+
+source=3
+target=9
+source_attribs= c(5,6,11,12,13)
+null_attrib=c(NA,NA,NA,NA,NA)
+name_attribs=c("app","location","hastag","lang","create at")
+name_file_csv<-"rstudio_RTs.csv"
+name_file_gdf<-"rstudio_RTs.gdf"
+directed <-TRUE
+```
+
+The sample CSV file can be found [here](/post/2019-06-18_files/rstudio_RTs.csv).
+
+## Define the data structures for the transformation
+
+In order to store the data, dynamic structures are needed to allow data to be added as they appear. The hash tables were chosen because they are the most appropriate for this case.
+Hash tables will be used to store nodes and connections.
+
+```r
+hash_nodes <- hash()
+hash_links <- hash()
+hash_links_in <- hash()
+hash_links_out <- hash()
+hash_connections <- hash()
+hash_connections_attrib<- hash()
+num_attribs<-length(source_attribs)
+```
+
+## Import the data from a file in CSV format and store the data in the structures
+
+Import data reading the CSV file and run it row by row to store the nodes and connections in the hash tables.
+
+Related entities can appear multiple times, as a source or as a target. When an entity appears for the first time, it is stored in the hash_nodes table. Attributes are associated to source entities and null attributes to target entities. It is a criterion that assumes this algorithm, but there could be others. If an entity appears first as a target as source, the null attributes are replaced by theirs as source.
```

Comment:
El siguiente comentario es un poco subjetivo, por lo que no es obligatorio el cambio, pero sí que convendría que lo reconsiderases.

La última frase: "if an entity appears...". Creo que no la llego a entender del todo (ojo: igual es culpa mía, de ahí lo subjetivo del comentario). Entiendo que dices que si un "entity" aparece como target cuando en verdad es source, luego se cambia lo del null attribute por su atributo no?

Si es así, creo que no estaría mal reformularla un poco para que hacerla un pelín más clara.

Ej: If at first, an entity appears as a target when it is actually a source, the null attribute assigned because it is considered a target, it will later be changed to the real one by the algorithm.

Reply:
Es un poco lioso de explicar. Las entidades pueden ser unas veces entidad fuente (hace un RT) o entidad destino (recibe un RT). Si aparece la primera vez como destino, se le asignarán los atributos nulos, pero si vuelve a aparecer como origen hay que sustituir los atributos nulos por los suyos.
He puesto esto que no sé si lo aclara mejor
If an entity appears the first time as a target, it will be assigned the null attributes, but if it appears later as a source, the null attributes will be replaced by theirs.